store/cma_store_test.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCMAInit(t *testing.T) {
+	store := NewCMAStore()
+	data := store.Get(time.Now().Add(-time.Minute), time.Now())
+	assert.Empty(t, len(data), time.Now())
+}
+
+func TestCMANilInsert(t *testing.T) {
+	store := NewCMAStore()
+	assert.Error(t, store.Put(TimePoint{time.Now(), nil}))
+}
+
+func TestCMAInsertNormal(t *testing.T) {
+	store := NewCMAStore()
+	now := time.Now()
+	assert := assert.New(t)
+	assert.NoError(store.Put(TimePoint{now, uint64(2)}))
+	assert.NoError(store.Put(TimePoint{now.Add(-time.Second), uint64(1)}))
+	assert.NoError(store.Put(TimePoint{now.Add(time.Second), uint64(3)}))
+	assert.NoError(store.Put(TimePoint{now.Add(-2 * time.Second), uint64(0)}))
+	actual := store.Get(time.Time{}, now)
+	require.Len(t, actual, 3)
+	actual = store.Get(time.Time{}, time.Time{})
+	require.Len(t, actual, 4)
+}
+
+func TestCMAInsertTime(t *testing.T) {
+	zeroTime := time.Time{}
+	store := NewCMAStore()
+	now := time.Now()
+	assert := assert.New(t)
+
+	// Errors
+	assert.Error(store.Put(TimePoint{zeroTime, uint64(2)}))
+	assert.Error(store.Put(TimePoint{zeroTime, nil}))
+
+	assert.NoError(store.Put(TimePoint{now, uint64(2)}))
+	assert.NoError(store.Put(TimePoint{now, uint64(6)}))
+	assert.NoError(store.Put(TimePoint{now.Add(2 * time.Second), uint64(0)}))
+	assert.NoError(store.Put(TimePoint{now.Add(2 * time.Second), uint64(20)}))
+	assert.NoError(store.Put(TimePoint{now.Add(2 * time.Second), uint64(10)}))
+	assert.NoError(store.Put(TimePoint{now.Add(2 * time.Second), uint64(1000)}))
+
+	// Get Invocation with values 2 and 4
+	actual := store.Get(zeroTime, now)
+	fmt.Println(actual)
+	require.Len(t, actual, 1)
+	assert.Equal(actual[0].Value, uint64(4))
+
+	// Get Invocation with values 0, 20, 10 and 1000
+	actual = store.Get(zeroTime, zeroTime)
+	require.Len(t, actual, 2)
+	assert.Equal(actual[0].Value, uint64(257)) // 257.5 rounded down
```

Comment:
The problem in that case is that (0+30+10)/3 is not round, as well as 1040/5, 1040/6 etc. As this is a cumulative average with unsigned values, the number of testcases that will actually match the exact average of all values is very limited, due to consecutive rounding for each individual average calculation. Even if we find such a testcase, that behavior is part of the the uint64 implementation and should require testing.